mlir/lib/Target/LLVMIR/ModuleTranslation.cpp

```diff
@@ -632,7 +632,7 @@ llvm::Constant *mlir::LLVM::detail::getLLVMConstant(
           llvm::ElementCount::get(numElements, /*Scalable=*/isScalable), child);
     if (llvmType->isArrayTy()) {
       auto *arrayType = llvm::ArrayType::get(elementType, numElements);
-      SmallVector<llvm::Constant *, 8> constants(numElements, child);
+      std::vector<llvm::Constant *> constants(numElements, child);
       return llvm::ConstantArray::get(arrayType, constants);
```

Comment:
We're in the splat case, we can have a fast path for zero by getting a `ConstantAggregateZero` instead here.

And if not, we can go the `ConstantDataArray` route instead when the element type is scalar.

Reply:
👍  Using `ConstantAggregateZero` improves things a lot. I updated the code to use it when constant is known to be zero. I reverted the std::vector to SmallVector since I don't really have a use case that would require it now.

Reply:
A splat of non-zero would have the problem, wouldn't it?

Reply:
I guess. I'll have a look at that as well.

Reply:
Is there a efficient way to represent splat constant arrays with a non-zero value in LLVM?

Reply:
That's a good question. This will still take long time to translate. 

```
llvm.mlir.global common @c_(dense<1> : vector<4294967296xi8>) {addr_space = 0 : i32} : !llvm.array<4294967296 x i8>
```

Reply:
Did you see the second part of my comment when I opened this thread?

> And if not, we can go the ConstantDataArray route instead when the element type is scalar.

Have you looked at `ConstantDataArray` ? It won't be a splat, but still hopefully better?

Reply:
As I can see we still need to pass a vector of the number of elements. Let me check if it improves anything.

Reply:
`ConstantArray` creation for scalar types will have to build a vector of the individual integers by extracting from the Constant object. So you're necessarily incurring more work in this case, see the code here for example: https://github.com/llvm/llvm-project/blob/53b904823cba4e93d58793b0804407f812477724/llvm/lib/IR/Constants.cpp#L1220-L1226

Now something a bit strange is that this is using a SmallVector as well, so it should fail the exact same way shouldn't it?

Reply:
Maybe the size limitation is not checked when using push_back?

Reply:
I just tested it and it fails with the same error if you don't initialize the SmallVector and then push child to it in a loop.

Reply:
Ah it pushes the integer to the `Elts` vector and not the Constant*. The integer is an i8 which is below 4 bytes so it uses a 64-bits size limit while for the Constant* the size is limited to 32-bits. I guess things would stop working if you use an i64 instead of an i8?

Reply:
That makes sense.

Reply:
@joker-eph Did you have time to have a look at the ConstantDataArray path?